scripts/upgrade_schema/upgrade_schema-59.X

```diff
@@ -0,0 +1,57 @@
+CREATE TABLE `seq_component` (
+  `id_seq_component`    BIGINT(20) UNSIGNED NOT NULL AUTO_INCREMENT COMMENT 'Auto-generated primary key',
+  `id_run`              BIGINT(20) UNSIGNED NOT NULL COMMENT 'Numeric run id',
+  `position`            TINYINT(1) UNSIGNED NOT NULL COMMENT 'Numeric position (lane number)',
+  `tag_index`           MEDIUMINT UNSIGNED  DEFAULT NULL COMMENT 'An optional numeric tag index',
+  `subset`              VARCHAR(10)         DEFAULT NULL COMMENT 'An optional sequence subset',
+  `digest`              CHAR(64) NOT NULL COMMENT 'A SHA256 hex digest of the component JSON representation as defined in npg_tracking::glossary::composition::component',
+  PRIMARY KEY (`id_seq_component`),
+  UNIQUE KEY `unq_seq_compon_d` (`digest`),
+  INDEX `unq_seq_compos_rp` (`id_run`, `position`, `tag_index`, `subset`)
+) ENGINE=InnoDB  DEFAULT CHARSET=latin1;
+
+CREATE TABLE `seq_composition` (
+  `id_seq_composition` BIGINT(20) UNSIGNED NOT NULL AUTO_INCREMENT COMMENT 'Auto-generated primary key',
+  `digest`             CHAR(64) NOT NULL COMMENT 'A SHA256 hex digest of the JSON representation of the composition as defined in npg_tracking::glossary::composition',
+  `size`               TINYINT UNSIGNED NOT NULL COMMENT 'Total number of components in a composition',
+  PRIMARY KEY (`id_seq_composition`),
+  UNIQUE KEY `unq_seq_compos_d` (`digest`),
+  UNIQUE KEY `unq_seq_compos_ps` (`id_seq_composition`, `size`) /* To allow for a composite foreign key in seq_component_composition */
+) ENGINE=InnoDB  DEFAULT CHARSET=latin1;
+
+CREATE TABLE `seq_component_composition` (
+  `id_seq_comcom`      BIGINT(20) UNSIGNED NOT NULL AUTO_INCREMENT COMMENT 'Auto-generated primary key',
+  `id_seq_component`   BIGINT(20) UNSIGNED NOT NULL COMMENT 'A foreign key referencing the id_seq_component column of the seq_component table',
+  `id_seq_composition` BIGINT(20) UNSIGNED NOT NULL COMMENT 'A foreign key referencing the id_seq_composition column of the seq_composition table',
+  `size`               TINYINT UNSIGNED NOT NULL    COMMENT 'Total number of components in a composition',
+  PRIMARY KEY (`id_seq_comcom`),
+  UNIQUE KEY `unq_seq_comcom` (`id_seq_component`, `id_seq_composition`, `size`),
```

Comment:
On 11/09/2015 12:12, dkj wrote:
>
> In scripts/upgrade_schema/upgrade_schema-59.X 
> <https://github.com/wtsi-npg/npg_qc/pull/216#discussion_r39261660>:
>
> > +CREATE TABLE `seq_composition` (
> > +  `id_seq_composition` BIGINT(20) UNSIGNED NOT NULL AUTO_INCREMENT COMMENT 'Auto-generated primary key',
> > +  `digest`             CHAR(64) NOT NULL COMMENT 'A SHA256 hex digest of the JSON representation of the composition as defined in npg_tracking::glossary::composition',
> > +  `size`               TINYINT UNSIGNED NOT NULL COMMENT 'Total number of components in a composition',
> > +  PRIMARY KEY (`id_seq_composition`),
> > +  UNIQUE KEY `unq_seq_compos_d` (`digest`),
> > +  UNIQUE KEY `unq_seq_compos_ps` (`id_seq_composition`, `size`) /* To allow for a composite foreign key in seq_component_composition */
> > +) ENGINE=InnoDB  DEFAULT CHARSET=latin1;
> > +
> > +CREATE TABLE `seq_component_composition` (
> > +  `id_seq_comcom`      BIGINT(20) UNSIGNED NOT NULL AUTO_INCREMENT COMMENT 'Auto-generated primary key',
> > +  `id_seq_component`   BIGINT(20) UNSIGNED NOT NULL COMMENT 'A foreign key referencing the id_seq_component column of the seq_component table',
> > +  `id_seq_composition` BIGINT(20cmpos. ) UNSIGNED NOT NULL COMMENT 'A foreign key referencing the id_seq_composition column of the seq_composition table',
> > +  `size`               TINYINT UNSIGNED NOT NULL    COMMENT 'Total number of components in a composition',
> > +  PRIMARY KEY (`id_seq_comcom`),
> > +  UNIQUE KEY `unq_seq_comcom` (`id_seq_component`, `id_seq_composition`, `size`),
>
> Does that mean |
> UNIQUE KEY `unq_seq_compos_ps` (`id_seq_composition`, `size`)
> | should lose size as well? (Else I've misunderstood its comment)
>

I think it is needed. Originally I had size in the linking table only. 
However, size is a property of the composition, not of the relationship 
of the individual component and composition. Therefore, I created the 
size column in the composition table. It is convenient to have it on a 
relationship as well as long as we can validate that the size value in 
the linking table is the same as in the composition table. The 
validation is set up by having a composite foreign key. In MySQL there 
are restrictions on what the composite key refers to in the parent 
table. Should be a primary key or a unique key (make sense). Hence the 
unique key on id_seq_composition and size in seq_composition.  It is not 
needed in the table itself, but, as far as I can see,  it does not 
contradict to anything in this table either.

Alternatively, if this feels too clunky, we can remove size from the 
linking table.

> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/wtsi-npg/npg_qc/pull/216/files#r39261660>.
>




-- 
 The Wellcome Trust Sanger Institute is operated by Genome Research 
 Limited, a charity registered in England with number 1021457 and a 
 company registered in England with number 2742969, whose registered 
 office is 215 Euston Road, London, NW1 2BE.